contracts/access/TimelockController.sol

```diff
@@ -0,0 +1,294 @@
+        return keccak256(abi.encode(targets, values, datas, predecessor, salt));
+    }
+
+    /**
+     * @dev Schedule an operation containing a single transaction.
+     *
+     * Emits a {CallScheduled} event.
+     *
+     * Requirements:
+     *
+     * - the caller must have the 'proposer' role.
+     */
+    function schedule(address target, uint256 value, bytes calldata data, bytes32 predecessor, bytes32 salt, uint256 delay) public onlyRole(PROPOSER_ROLE) {
+        bytes32 id = hashOperation(target, value, data, predecessor, salt);
+        _schedule(id, delay);
+        emit CallScheduled(id, 0, target, value, data, predecessor);
+    }
+
+    /**
+     * @dev Schedule an operation containing a batch of transactions.
+     *
+     * Emits one {CallScheduled} event per transaction in the batch.
+     *
+     * Requirements:
+     *
+     * - the caller must have the 'proposer' role.
+     */
+    function scheduleBatch(address[] calldata targets, uint256[] calldata values, bytes[] calldata datas, bytes32 predecessor, bytes32 salt, uint256 delay) public onlyRole(PROPOSER_ROLE) {
+        require(targets.length == values.length, "TimelockController: length missmatch");
+        require(targets.length == datas.length, "TimelockController: length missmatch");
+
+        bytes32 id = hashOperationBatch(targets, values, datas, predecessor, salt);
+        _schedule(id, delay);
+        for (uint256 i = 0; i < targets.length; ++i) {
+            emit CallScheduled(id, i, targets[i], values[i], datas[i], predecessor);
+        }
+    }
+
+    /**
+     * @dev Schedule an operation that is to becomes valid after a given delay.
+     */
+    function _schedule(bytes32 id, uint256 delay) internal {
+        require(_timestamps[id] == 0, "TimelockController: operation already scheduled");
+        require(delay >= _minDelay, "TimelockController: insufficient delay");
+        // solhint-disable-next-line not-rely-on-time
+        _timestamps[id] = SafeMath.add(block.timestamp, delay);
+    }
+
+    /**
+     * @dev Cancel an operation.
+     *
+     * Requirements:
+     *
+     * - the caller must have the 'proposer' role.
+     */
+    function cancel(bytes32 id) public onlyRole(PROPOSER_ROLE) {
+        require(!isOperationDone(id), "TimelockController: operation is already executed");
+        delete _timestamps[id];
+
+        emit Cancelled(id);
+    }
+
+    /**
+     * @dev Execute an (ready) operation containing a single transaction.
+     *
+     * Emits a {CallExecuted} event.
+     *
+     * Requirements:
+     *
+     * - the caller must have the 'executor' role.
+     */
+    function execute(address target, uint256 value, bytes calldata data, bytes32 predecessor, bytes32 salt) public payable onlyRole(EXECUTOR_ROLE) {
+        bytes32 id = hashOperation(target, value, data, predecessor, salt);
+        _execute(id, predecessor);
```

Comment:
The operation should be marked as executed only after all calls are made, as protection against potential reentrancy.

Reply:
Reentrancy trying to cancel the current operations are not an issue since the would cause the tx to revert, regarding of the order.

Similarly, an operation trying to re-execute itself would lead to a revert. If the marking is done before the call, the "inner" execute would revert, which might eventually get caught. If the marking is done after the call, the "outer" execute would revert, thus reverting everything. So for re-executing reentrancy, I'd prefer marking before running.

There is a reentrancy issue if a batch operation tries to execute its child before all the call in the batch have been called. This is possible since the batch would be marked as done before all calls have been done. This is potentially dangerous but is very difficult to do:
- The father operation needs to include the execution of the child operation, so it must be part of the data that produces the id
- The child operation's id depend on its predecessor's ...
This hash loop can be broken by storing the child operation id in an execution proxy that would be called by the father operation.

Have I missed other reentrancy attacks?

Reply:
Drawback with checking after the call is that is the check is to fail, gas has already been consumed

Reply:
What about only checking before the call, then re-checking and marking as executed after the call?

Yes I think the only danger is with batch operations as you mentioned. It's true that an attack would be difficult to put together but I don't think it justifies not putting in place the proper protections.

Reply:
Double-checking saves gas for reverts, but increases for successful operations :/

Last commits include:
1. Checking the predecessor before doing the call
2. Checking pre-execution after the call
3. Marking completion after the call

1+3 prevent ordering error (particularly with batch)
2+3 prevents reentrancy on the same operation

Reply:
Perfect!